# PySpark en Google Colab

Instalacion Octubre/2021

1.   Instalacion Java
2.   Instalacion de Spark
3.   Instalar PySpark

## De forma General para usar pyspark en Colab (2021) siga los siguientes pasos en una celda en Colab:

In [16]:
# instalar Java
!apt-get install openjdk-8-jdk-headless -qq > /dev/null


In [17]:
# Descargar la ultima versión de java ( comprobar que existen los path de descarga)
# Download latest release. Update if necessary

!wget -q https://downloads.apache.org/spark/spark-3.1.2/spark-3.1.2-bin-hadoop2.7.tgz

In [18]:
%ls -la /content/



total 438396
drwxr-xr-x  1 root root      4096 Oct  5 07:52 ./
drwxr-xr-x  1 root root      4096 Oct  5 06:54 ../
drwxr-xr-x  4 root root      4096 Sep 30 17:11 .config/
drwxr-xr-x  1 root root      4096 Sep 30 17:12 sample_data/
drwxr-xr-x 13 1000 1000      4096 May 24 05:00 spark-3.1.2-bin-hadoop2.7/
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz.1


In [19]:
!tar xf spark-3.1.2-bin-hadoop2.7.tgz

In [20]:
%ls -la

total 438396
drwxr-xr-x  1 root root      4096 Oct  5 07:52 ./
drwxr-xr-x  1 root root      4096 Oct  5 06:54 ../
drwxr-xr-x  4 root root      4096 Sep 30 17:11 .config/
drwxr-xr-x  1 root root      4096 Sep 30 17:12 sample_data/
drwxr-xr-x 13 1000 1000      4096 May 24 05:00 spark-3.1.2-bin-hadoop2.7/
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz.1


In [21]:
# instalar pyspark
!pip install -q pyspark

# Variables de entorno

In [22]:
import os # libreria de manejo del sistema operativo
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.1.2-bin-hadoop2.7"

# Cargar pyspark en el sistema

In [23]:
from pyspark.sql import SparkSession

APP_NAME = "PDGE-tutorialSpark1"
SPARK_URL = "local[*]"
spark = SparkSession.builder.appName(APP_NAME).master(SPARK_URL).getOrCreate()
spark

Pruebas


In [24]:
sc = spark.sparkContext
#obtener el contexto de ejecución de Spark del Driver.
sc

<SparkContext master=local[*] appName=PDGE-tutorialSpark1>

In [25]:
array = sc.parallelize([1,2,3,4,5,6,7,8,9,10], 2)
array

ParallelCollectionRDD[4] at readRDDFromFile at PythonRDD.scala:274

In [26]:
print(array.collect())

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [27]:
print(array.count())


10


In [28]:
num3 = array.map(lambda elemento: 3*elemento)

In [29]:
print(num3.collect())

[3, 6, 9, 12, 15, 18, 21, 24, 27, 30]


# Ejemplos de operaciones con RDDs de Spark

In [30]:
numeros = sc.parallelize([1,2,3,4,5,6,7,8,9,10],2)

print(numeros.reduce(lambda e1,e2: e1+e2))

55


In [31]:
pnumeros = sc.parallelize([1,2,3,4,5,6,7,8,9,10])

rdd = pnumeros.map(lambda e: 2*e)

print(rdd.collect())


[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


## Pregunta TS1.1 ¿Cómo hacer para obtener una lista de los elementos al cuadrado?

In [32]:
#Mapeamos un lambda 
pnumeros = sc.parallelize([1,2,3,4,5,6,7,8,9,10])

rdd = pnumeros.map(lambda e: e**2)

print(rdd.collect())

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [33]:
numeros = sc.parallelize([1,2,3,4,5,6,7,8,9,10])

rdd = numeros.filter(lambda elemento: elemento%2==0)

print (rdd.collect())

[2, 4, 6, 8, 10]


## Pregunta TS1.2 ¿Cómo filtrar los impares?

In [34]:
rddi = numeros.filter (lambda e: e%2==1)
print (rddi.collect())

[1, 3, 5, 7, 9]


#Operaciones con números

Sumar n numeros

In [35]:
rddnum=sc.range(1000000)
rddnum.reduce(lambda a,b: a+b)

499999500000

In [36]:
numeros = sc.parallelize([1,1,2,2,5])

unicos = numeros.distinct()

print (unicos.collect())

[2, 1, 5]


In [37]:
numeros = sc.parallelize([1,2,3,4,5])

rdd = numeros.flatMap(lambda elemento: [elemento, 10*elemento])

print (rdd.collect())

[1, 10, 2, 20, 3, 30, 4, 40, 5, 50]


Muestreo

In [38]:
numeros = sc.parallelize([1,2,3,4,5,6,7,8,9,10])

rdd = numeros.sample(True, 1.0)
# True = sin reemplazar
# 1.0 = fracción de elementos = 100%
print (rdd.collect())

[1, 2, 2, 3, 3, 5, 5, 6, 7, 8, 8, 8, 9]


Union 

In [39]:
pares = sc.parallelize([2,4,6,8,10])
impares = sc.parallelize([1,3,5,7,9])

numeros = pares.union(impares)

print (numeros.collect())

[2, 4, 6, 8, 10, 1, 3, 5, 7, 9]


reducción


In [40]:
numeros = sc.parallelize([1,2,3,4,5])
# operación asociativa y conmutativa
# orden no está definido
print (numeros.reduce(lambda elem1,elem2: elem2+elem1))

15


## Pregunta TS1.3 ¿Tiene sentido esta operación? ¿Si se repite se obtiene siempre el mismo resultado?

No, no tiene sentido ya que la función pasada a reduce debe de ser asociativa y conmutativa. En este caso es una resta, que no tiene ninguna de las 2 propiedades. Dado que en reduce el orden no está definido, al parelelizar, cada vez que hacemos la operación obtenemos resultados distintos

In [41]:
#Tiene sentido esta operación?
numeros = sc.parallelize([1,2,3,4,5])

print (numeros.reduce(lambda elem1,elem2: elem1-elem2))

5


In [42]:
#Tiene sentido esta operación?
numeros = sc.parallelize([1,2,4,3,5])

print (numeros.reduce(lambda elem1,elem2: elem1-elem2))

3


#Acciones

In [43]:
numeros = sc.parallelize([5,3,2,1,4])

print (numeros.take(3))
#¿Qué sucede si ponemos 30 en vez de 3 elementos?

[5, 3, 2]


Que obtenemos el mismo resultado ya que take solo toma los n primeros elementos, en este caso los 3 primeros

In [44]:
numeros = sc.parallelize([3,2,1,4,5])
# 
print (numeros.takeOrdered(3, lambda elem: -elem))
# La función lambda se está utilizando para crear el índice de la lista de ordenación 

[5, 4, 3]


## Pregunta TS1.4 ¿Cómo lo ordenarías para que primero aparezcan los impares y luego los pares?

In [45]:
numeros = sc.parallelize([1,2,3,4,5,6,7,8,9,10])
pares = numeros.filter(lambda e: e%2==0)
impares = numeros.filter(lambda e: e%2==1)
nums_ordenados_impar_par = impares.union(pares)

nums_ordenados_impar_par.collect()

[1, 3, 5, 7, 9, 2, 4, 6, 8, 10]

# RDDs con string

In [46]:
palabras = sc.parallelize(['HOLA', 'Que', 'TAL', 'Bien'])

pal_minus = palabras.map(lambda elemento: elemento.lower())

print (pal_minus.collect())

['hola', 'que', 'tal', 'bien']


In [47]:
palabras = sc.parallelize(['HOLA', 'Que', 'TAL', 'Bien'])

pal_long = palabras.map(lambda elemento: len(elemento))

print (pal_long.collect())

[4, 3, 3, 4]


In [48]:
log = sc.parallelize(['E: e21', 'W: w12', 'W: w13', 'E: e45'])

errors = log.filter(lambda elemento: elemento[0]=='E')

print (errors.collect())

['E: e21', 'E: e45']


In [49]:
lineas = sc.parallelize(['', 'a', 'a b', 'a b c'])

palabras = lineas.flatMap(lambda elemento: elemento.split())

print (palabras.collect())

['a', 'a', 'b', 'a', 'b', 'c']


In [50]:
lineas = sc.parallelize(['', 'a', 'a b', 'a b c'])

palabras_flat = lineas.flatMap(lambda elemento: elemento.split())
palabras_map = lineas.map(lambda elemento: elemento.split())

print (palabras_flat.collect())
print (palabras_map.collect())


['a', 'a', 'b', 'a', 'b', 'c']
[[], ['a'], ['a', 'b'], ['a', 'b', 'c']]


## Pregunta TS1.5 ¿Cuántos elementos tiene cada rdd? ¿Cuál tiene más?


El primero tiene 6 elementos, el segundo 4(ya que estamos aplicando lamba a cada elemento y cada elemento produce una salida). En el primero aplicamos flatmap, donde cada elemento puede producir más de una salida, por eso el primer rdd tiene más elementos

## Pregunta TS1.6 ¿De qué tipo son los elementos del rdd palabras_map? ¿Por qué palabras_map tiene el primer elemento vacío?

Son listas. Porque al aplicar la función split sobre un string vacio no obtenemos ningun elemento(es decir, una lista vacía)

In [51]:
palabras_flat.take(1)

['a']

In [52]:
palabras_map


PythonRDD[56] at collect at <ipython-input-50-9e9e885b2c3e>:7

## Pregunta TS1.7. Prueba la transformación distinct si lo aplicamos a cadenas.

In [53]:
rep_info = sc.parallelize(['a', 'A','b', 'b', 'c', 'd','e','f','f'])
rdd_distinct_info = rep_info.distinct()
print(rdd_distinct_info.collect())

['b', 'c', 'd', 'a', 'A', 'e', 'f']


In [54]:
log = sc.parallelize(['E: e21', 'I: i11', 'W: w12', 'I: i11', 'W: w13', 'E: e45'])

infos = log.filter(lambda elemento: elemento[0]=='I')
errors = log.filter(lambda elemento: elemento[0]=='E')

inferr = infos.union(errors)

print (inferr.collect())

['I: i11', 'I: i11', 'E: e21', 'E: e45']


## Pregunta TS1.8 ¿Cómo se podría obtener la misma salida pero utilizando una sola transformación y sin realizar la unión?

Añadiendo un or a la lambda por ejemplo:

In [55]:
log = sc.parallelize(['E: e21', 'I: i11', 'W: w12', 'I: i11', 'W: w13', 'E: e45'])

infos_and_errors = log.filter(lambda elemento: elemento[0]=='I' or elemento[0]=='E' )

print (infos_and_errors.collect())

['E: e21', 'I: i11', 'I: i11', 'E: e45']


In [56]:
numeros = sc.parallelize([1,2,3,4,5])

print (numeros.reduce(lambda elem1,elem2: elem2+elem1))

15


In [57]:
#Tiene sentido esta operación?
numeros = sc.parallelize([1,2,3,4,5])

print (numeros.reduce(lambda elem1,elem2: elem1-elem2))

5


No, la resta no es asociativa ni conmutativa para meterla en reduce

In [58]:
palabras = sc.parallelize(['HOLA', 'Que', 'TAL', 'Bien'])

pal_minus = palabras.map(lambda elemento: elemento.lower())

print (pal_minus.reduce(lambda elem1,elem2: elem1+"-"+elem2))
#y esta tiene sentido esta operación?
# Qué pasa si ponemos elem2+"-"+elem1

hola-que-tal-bien


RESPUESTA: 
  Si queremos que siempre de la misma salida: "hola-que-tal-bien" no, ya que esa lambda no es conmuntativa

In [59]:
print (pal_minus.reduce(lambda elem1,elem2: elem2+"-"+elem1))


bien-tal-que-hola


In [60]:
r = sc.parallelize([('A', 1),('C', 4),('A', 1),('B', 1),('B', 4)])
rr = r.reduceByKey(lambda v1,v2:v1+v2)
print (rr.collect())

[('C', 4), ('A', 2), ('B', 5)]


In [61]:
r = sc.parallelize([('A', 1),('C', 4),('A', 1),('B', 1),('B', 4)])
rr1 = r.reduceByKey(lambda v1,v2:v1+v2)
print (rr1.collect())
rr2 = rr1.reduceByKey(lambda v1,v2:v1)
print (rr2.collect())


[('C', 4), ('A', 2), ('B', 5)]
[('C', 4), ('A', 2), ('B', 5)]


## Pregunta TS1.9 ¿Cómo explica el funcionamiento de las celdas anteriores?

rr2 no hace reduceByKey de r (como rr1), si que que hace reduceByKey de rr1, donde todos los elementos tienen clave única, por lo que realmente no está aplicando nada y obtenemos la misma salida

In [62]:
r = sc.parallelize([('A', 1),('C', 4),('A', 1),('B', 1),('B', 4)])
rr1 = r.reduceByKey(lambda v1,v2:'hola')
print (rr1.collect())
rr2 = rr1.reduceByKey(lambda v1,v2:'hola')
print (rr2.collect())

[('C', 4), ('A', 'hola'), ('B', 'hola')]
[('C', 4), ('A', 'hola'), ('B', 'hola')]


In [63]:
r = sc.parallelize([('A', 1),('C', 2),('A', 3),('B', 4),('B', 5)])
rr = r.groupByKey()
res= rr.collect()
for k,v in res:
    print (k, list(v))
# Que operación realizar al RDD rr para que la operacion sea como un reduceByKey
#¿Y simular un group con un reduceByKey y un map?

C [2]
A [1, 3]
B [4, 5]


In [64]:
rdd1 = sc.parallelize([('A',1),('B',2),('C',3)])
rdd2 = sc.parallelize([('A',4),('B',5),('C',6)])

rddjoin = rdd1.join(rdd2)

print (rddjoin.collect())
# Prueba a cambiar las claves del rdd1 y rdd2 para ver cuántos elementos se crean

[('A', (1, 4)), ('B', (2, 5)), ('C', (3, 6))]


In [65]:
rdd1 = sc.parallelize([('A',1),('B',2),('C',3)])
rdd2 = sc.parallelize([('A',4),('A',5),('B',6),('D',7)])

rddjoin = rdd1.join(rdd2)

print (rddjoin.collect())
#Modifica join por leftOuterJoin, rightOuterJoin y fullOuterJoin ¿Qué sucede? 


[('A', (1, 4)), ('A', (1, 5)), ('B', (2, 6))]


Con leftOuterJoin: [('A', (1, 4)), ('A', (1, 5)), ('B', (2, 6)), ('C', (3, None))]

Con rightOuterJoin: [('A', (1, 4)), ('A', (1, 5)), ('B', (2, 6)), ('D', (None, 7))]

Con fullOuterJoin : [('A', (1, 4)), ('A', (1, 5)), ('B', (2, 6)), ('C', (3, None)), ('D', (None, 7))]

Es decir, con el left obliga a combinar todas las claves de la izquierda, aunque no estén en el lado derecho, por eso la entrada None en la clave C.
De manera similar ocurre con el right y la Clave D.
Con el full en obtenemos las claves C y D, es decir, obligamos a combinar las claves de cada lado, si están en un lado y en el otro no, aparecerán con un valor None.


In [66]:
rdd = sc.parallelize([('A',1),('B',2),('C',3),('A',4),('A',5),('B',6)])

res = rdd.sortByKey(False)

print (res.collect())

[('C', 3), ('B', 2), ('B', 6), ('A', 1), ('A', 4), ('A', 5)]


# Utilización de ficheros

In [67]:

# Crea una lista de 1000 números y la guarda.
# Da error si 'salida' existe, descomenta la linea de borrar en ese caso.
%rm -rf salida
#numeros = sc.parallelize(range(0,1000))
numeros = sc.parallelize(range(0,1000),8)


numeros.saveAsTextFile('salida')


In [68]:
ls -la

total 438400
drwxr-xr-x  1 root root      4096 Oct  5 07:53 ./
drwxr-xr-x  1 root root      4096 Oct  5 06:54 ../
drwxr-xr-x  4 root root      4096 Sep 30 17:11 .config/
drwxr-xr-x  2 root root      4096 Oct  5 07:53 salida/
drwxr-xr-x  1 root root      4096 Sep 30 17:12 sample_data/
drwxr-xr-x 13 1000 1000      4096 May 24 05:00 spark-3.1.2-bin-hadoop2.7/
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz.1


In [69]:
%ls -la salida/*


-rw-r--r-- 1 root root 390 Oct  5 07:53 salida/part-00000
-rw-r--r-- 1 root root 500 Oct  5 07:53 salida/part-00001
-rw-r--r-- 1 root root 500 Oct  5 07:53 salida/part-00002
-rw-r--r-- 1 root root 500 Oct  5 07:53 salida/part-00003
-rw-r--r-- 1 root root 500 Oct  5 07:53 salida/part-00004
-rw-r--r-- 1 root root 500 Oct  5 07:53 salida/part-00005
-rw-r--r-- 1 root root 500 Oct  5 07:53 salida/part-00006
-rw-r--r-- 1 root root 500 Oct  5 07:53 salida/part-00007
-rw-r--r-- 1 root root   0 Oct  5 07:53 salida/_SUCCESS


In [70]:
%cat salida/part-00000

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124


In [71]:
# Recupera el fichero guardado y realiza la suma
n2 = sc.textFile('salida').map(lambda a:int(a))
print(n2.reduce(lambda v1,v2: v1 + v2))

# Prueba este código y mira qué genera?
# Borra la salida y cambia las particiones en parallelize ¿Qué sucede? 
#    (pe c.parallelize(xrange(0,1000),8))


499500


## Pregunta TS1.10 Borra la salida y cambia las particiones en parallelize ¿Qué sucede? 
  (pe c.parallelize(xrange(0,1000),8))


In [72]:
%rm -rf salida
%ls -la


total 438396
drwxr-xr-x  1 root root      4096 Oct  5 07:53 ./
drwxr-xr-x  1 root root      4096 Oct  5 06:54 ../
drwxr-xr-x  4 root root      4096 Sep 30 17:11 .config/
drwxr-xr-x  1 root root      4096 Sep 30 17:12 sample_data/
drwxr-xr-x 13 1000 1000      4096 May 24 05:00 spark-3.1.2-bin-hadoop2.7/
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz.1


Que en salida obtenemos 8 particiones del RDD, desde part-00000 a part-00007 y ahora part-00000 solo contiene los primeros 125 numeros(hasta 124).


# El quijote

Montar el directorio de trabajo utilizando Google Drive
Loading Your Data into Google Colaboratory.
1. First of all, Upload your Data to your Google Drive.
2. Run the following script in colab shell.
3. Copy the authorization code of your account.
4. Paste the authorization code into the output shell.
5. Congrats! Now your Google Drive is mounted to this location /content/gdrive/MyDrive/




In [73]:
from google.colab import drive
#drive.flush_and_unmount()
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [74]:
%ls /content/gdrive/MyDrive/PDGE/

elquijote.txt


In [75]:
!cp "/content/gdrive/My Drive/PDGE/elquijote.txt" .

In [76]:
%ls -la

total 438712
drwxr-xr-x  1 root root      4096 Oct  5 07:58 ./
drwxr-xr-x  1 root root      4096 Oct  5 06:54 ../
drwxr-xr-x  4 root root      4096 Sep 30 17:11 .config/
-rw-------  1 root root    317618 Oct  5 07:58 elquijote.txt
drwx------  5 root root      4096 Oct  5 07:58 gdrive/
drwxr-xr-x  1 root root      4096 Sep 30 17:12 sample_data/
drwxr-xr-x 13 1000 1000      4096 May 24 05:00 spark-3.1.2-bin-hadoop2.7/
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz
-rw-r--r--  1 root root 224445805 May 24 05:01 spark-3.1.2-bin-hadoop2.7.tgz.1


In [77]:
%pwd

'/content'

# Procesando el QUIJOTE

In [78]:
quijote = sc.textFile("elquijote.txt")
quijote.take(10)

['EL INGENIOSO HIDALGO DON QUIJOTE DE LA MANCHA',
 '',
 'Miguel de Cervantes Saavedra',
 '',
 '     Capítulo primero',
 '',
 '     Que trata de la condición y ejercicio del famoso hidalgo D.',
 'Quijote de la ',
 '     Mancha',
 '']

Transformaciones

In [79]:
charsPerLine = quijote.map(lambda s: len(s))
allWords = quijote.flatMap(lambda s: s.split())
allWordsNoArticles = allWords.filter(lambda a: a.lower() not in ["el", "la"])
allWordsUnique = allWords.map(lambda s: s.lower()).distinct()
sampleWords = allWords.sample(withReplacement=True, fraction=0.2, seed=666)
weirdSampling = sampleWords.union(allWordsNoArticles.sample(False, fraction=0.3))
# cómo funciona cada transformación

In [80]:
allWordsUnique.take(10)

['el',
 'hidalgo',
 'don',
 'mancha',
 'saavedra',
 'que',
 'condición',
 'y',
 'del',
 'd.']

## Pregunta TS2.1 Explica la utilidad de cada transformación y detalle para cada una de ellas si cambia el número de elementos en el RDD resultante. Es decir si el RDD de partida tiene N elementos, y el de salida M elementos, indica si N>M, N=M o N<M.

* map: N=M Devuelve un nuevo rdd pasandole una función func a cada elemento del dataset origen 
* flatmap: N<=M lo mismo que map pero cada entrada puede ser mapeada a varias salidas(tambien conocido como secuencia).
* filter: N>=M Devuelve un rdd con aquellos elementos del dataset cuya función pasada quede evaluada como verdadero.
* distinct: N>=M Devuelve un rdd cuyos elementos son unicos y no repetidos
* sample: N>=M Devuelve una fracción de los elementos del rdd origen. Con o sin remplazamiento y de manera aleatoria. Es posible pasarle una semilla.
* union: N<=M Devuelve un rdd con la union del dataset origen con el pasado por argumento.


In [81]:
numLines = quijote.count()
numChars = charsPerLine.reduce(lambda a,b: a+b) # also charsPerLine.sum()
sortedWordsByLength = allWordsNoArticles.takeOrdered(20, key=lambda x: -len(x))
numLines, numChars, sortedWordsByLength

(5534,
 305678,
 ['procuremos.Levántate,',
  'estrechísimamente,',
  'Pintiquiniestra,',
  'entretenimiento,',
  'maravillosamente',
  'descansadamente;',
  'desenfadadamente',
  'quebrantamientos',
  'quebrantamiento,',
  'alternativamente',
  'encantamientos,',
  'Placerdemivida,',
  'encantamientos;',
  'malbaratándolas',
  'regocijadamente',
  'consentimiento,',
  'desaconsejaban,',
  'acontecimiento.',
  'agradeciéndoles',
  'encantamientos,'])

## Pregunta TS2.2 Explica el funcionamiento de cada acción anterior 

Implementa la opción count de otra manera:


*   utilizando transformaciones map y reduce
*   utilizando solo reduce en caso de que sea posible.



### TS2.2 RESPUESTA:

1. En la primera linea contamos el nuemro de lineas que tiene el quijote

2. numChars: para calcular el numero de caracteres, totales utilizamos el rdd  anterior charsperline y le aplicamos una lambda para que nos cuente todos los conteos de cada linea individual para obtener el total de caracteres

3. sortedWordsByLength: De nuevo utilizando el rdd anterior allWordsNoArticles, donde tenemos todas las palabras, y ordenamos por claves (las palabras en si), de mayor a menor segun la longitud de las palabras. Es decir, obtenemos las 20 palabras más largas.

In [82]:
numLines_2_mapped = quijote.map(lambda a:1 )  

numLines_2 = numLines_2_mapped.reduce(lambda a,b: a+b )  
numLines_2

5534

# Operaciones K-V (Clave Valor)

In [83]:
import requests
import re
allWords = allWords.flatMap(lambda w: re.sub(""";|:|\.|,|-|–|"|'|\s"""," ", w.lower()).split(" ")).filter(lambda a: len(a)>0)
allWords2 = sc.parallelize(requests.get("https://gist.githubusercontent.com/jsdario/9d871ed773c81bf217f57d1db2d2503f/raw/585de69b0631c805dabc6280506717943b82ba4a/el_quijote_ii.txt").iter_lines())
allWords2 = allWords2.flatMap(lambda w: re.sub(""";|:|\.|,|-|–|"|'|\s"""," ", w.decode("utf8").lower()).split(" ")).filter(lambda a: len(a)>0)

allWords y allWords2 representan todas las palabras del quijote. allWords se ha obtenido del fichero local, allWords2 de un url con el archivo en remoto. 
Para obtener todas las palabras se ha aplicado un flatmap a cada linea, cuya funcion lambda consiste en un expresion regular que quita los espacios para ver los conjunto de "aparentes palabras" y luego filtrarlas para quitar los simbulos y signos de puntuacion

In [84]:
allWords.take(10)

['el',
 'ingenioso',
 'hidalgo',
 'don',
 'quijote',
 'de',
 'la',
 'mancha',
 'miguel',
 'de']

Tomamos 10 palabras del primer conjunto 

In [85]:
allWords2.take(10)

['don',
 'quijote',
 'de',
 'la',
 'mancha',
 'miguel',
 'de',
 'cervantes',
 'saavedra',
 'segunda']

Tomamos 10 palabras del segundo conjunto 

In [86]:
words = allWords.map(lambda e: (e,1))
words2 = allWords2.map(lambda e: (e,1))

words.take(10)

[('el', 1),
 ('ingenioso', 1),
 ('hidalgo', 1),
 ('don', 1),
 ('quijote', 1),
 ('de', 1),
 ('la', 1),
 ('mancha', 1),
 ('miguel', 1),
 ('de', 1)]

A cada palabra la convertirmos(en otro rdd) en un diccionario tal que (clave=palabra,valor=1)

In [87]:
frequencies = words.reduceByKey(lambda a,b: a+b)
frequencies2 = words2.reduceByKey(lambda a,b: a+b)
frequencies.takeOrdered(10, key=lambda a: -a[1])

[('que', 3032),
 ('de', 2809),
 ('y', 2573),
 ('a', 1426),
 ('la', 1423),
 ('el', 1232),
 ('en', 1155),
 ('no', 903),
 ('se', 753),
 ('los', 696)]

Ya que tenemos diccionarios con claves repetidas, siguiendo el modelo map reduce, hacemos un reducebykey para obtener la frecuencia de cada palabra

In [88]:
res = words.groupByKey().takeOrdered(10, key=lambda a: -len(a))
res # To see the content, res[i][1].data
#for i in range(0,10):
#  print(res[i][1].data)

# el: print(sum(res[0][1].data)) da 1232 como abajo

[('el', <pyspark.resultiterable.ResultIterable at 0x7ff36477d0d0>),
 ('hidalgo', <pyspark.resultiterable.ResultIterable at 0x7ff36477d2d0>),
 ('don', <pyspark.resultiterable.ResultIterable at 0x7ff36477d350>),
 ('mancha', <pyspark.resultiterable.ResultIterable at 0x7ff36477d1d0>),
 ('saavedra', <pyspark.resultiterable.ResultIterable at 0x7ff36477d090>),
 ('que', <pyspark.resultiterable.ResultIterable at 0x7ff36477dbd0>),
 ('condición', <pyspark.resultiterable.ResultIterable at 0x7ff36477dc10>),
 ('y', <pyspark.resultiterable.ResultIterable at 0x7ff36477de10>),
 ('del', <pyspark.resultiterable.ResultIterable at 0x7ff36477db10>),
 ('d', <pyspark.resultiterable.ResultIterable at 0x7ff36477dd50>)]

Agrupamos las palabras por clave para obtener la frecuencia y cogemos las 10 con más frecuencia

In [89]:
joinFreq = frequencies.join(frequencies2)
joinFreq.take(10)

[('el', (1232, 4394)),
 ('hidalgo', (14, 42)),
 ('don', (370, 1606)),
 ('mancha', (26, 101)),
 ('saavedra', (1, 1)),
 ('que', (3032, 10040)),
 ('y', (2573, 9650)),
 ('del', (415, 1344)),
 ('en', (1155, 4223)),
 ('cuyo', (11, 35))]

Hacemos un join de las frecuecias obtenidas por cada metodo descrito arriba y tomamos 10 ejemplos (join da  el siguiente formato de salida(K, (V, W))) siendo K la clave de unión y V y W valores  de las tablas

In [90]:
test = frequencies.filter(lambda e: e[0] == 'pieza')
t = test.collect()

print(t[0][1])


9


In [91]:
test2 = frequencies2.filter(lambda e: e[0] == 'pieza')
test2.collect()

[('pieza', 1)]

In [92]:
joinFreq.map(lambda e: (e[0], (e[1][0] - e[1][1])/(e[1][0] + e[1][1]))).takeOrdered(10, lambda v: -v[1]), joinFreq.map(lambda e: (e[0], (e[1][0] - e[1][1])/(e[1][0] + e[1][1]))).takeOrdered(10, lambda v: +v[1])

([('pieza', 0.8),
  ('corral', 0.8),
  ('rodela', 0.7777777777777778),
  ('curar', 0.75),
  ('valle', 0.75),
  ('entierro', 0.75),
  ('oh', 0.7142857142857143),
  ('licor', 0.7142857142857143),
  ('difunto', 0.7142857142857143),
  ('pago', 0.6666666666666666)],
 [('teresa', -0.9767441860465116),
  ('roque', -0.96),
  ('paje', -0.9565217391304348),
  ('duque', -0.9565217391304348),
  ('blanca', -0.9565217391304348),
  ('gobernador', -0.9503105590062112),
  ('diego', -0.9459459459459459),
  ('tarde', -0.9428571428571428),
  ('mesmo', -0.9381443298969072),
  ('letras', -0.9354838709677419)])

Para cada uno de los métodos frecuencies y frecuencies2 calcula cociente entre la diferencia de ambos resultados y la suma de ambos resultados.
Por ejemplo "pieza" en frecuencies nos da "9" y en frecuencies 2 da "1" por lo que 8/10 = 0.8
Todos los resultados los da ordenados de mayor a menor(tomando solo los 10 primero), esto es, el resultado de esta operación.

Repetimos el resultado solo que ahora con resultados negativos

## Pregunta TS2.3 Explica el proposito de cada una de las operaciones anteriores

(Ver cuadro de texto debajo de cada operacion)

## Pregunta TS2.4 ¿Cómo puede implementarse la frecuencia con groupByKey y transformaciones?

Como hemos realizado en los apartado anteriores. Cremos un conjunto clave valor (K=palabra,1) para cada palabra, y hacemos un groupbykey. Finalmente con la secuencia de groupbykey(de 1's) calculamos su longitud obteniendo el resultado(frecuencia).

In [93]:
#Partiendo del rdd "allWords"con todas las palabras 
words = allWords.map(lambda e: (e,1))
#frequencies = words.groupByKey().mapValues(len).collect() # to get all freqs
frequencies = words.groupByKey().mapValues(len).take(10)

frequencies



[('el', 1232),
 ('hidalgo', 14),
 ('don', 370),
 ('mancha', 26),
 ('saavedra', 1),
 ('que', 3032),
 ('condición', 16),
 ('y', 2573),
 ('del', 415),
 ('d', 14)]

# Optimizaciones


## Pregunta TS2.5 ¿Cuál de las dos siguientes celdas es más eficiente? Justifique la respuesta.

In [94]:
joinFreq.map(lambda e: (e[0], (e[1][0] - e[1][1])/(e[1][0] + e[1][1]))).takeOrdered(10, lambda v: -v[1]), joinFreq.map(lambda e: (e[0], (e[1][0] - e[1][1])/(e[1][0] + e[1][1]))).takeOrdered(10, lambda v: +v[1])

([('pieza', 0.8),
  ('corral', 0.8),
  ('rodela', 0.7777777777777778),
  ('curar', 0.75),
  ('valle', 0.75),
  ('entierro', 0.75),
  ('oh', 0.7142857142857143),
  ('licor', 0.7142857142857143),
  ('difunto', 0.7142857142857143),
  ('pago', 0.6666666666666666)],
 [('teresa', -0.9767441860465116),
  ('roque', -0.96),
  ('paje', -0.9565217391304348),
  ('duque', -0.9565217391304348),
  ('blanca', -0.9565217391304348),
  ('gobernador', -0.9503105590062112),
  ('diego', -0.9459459459459459),
  ('tarde', -0.9428571428571428),
  ('mesmo', -0.9381443298969072),
  ('letras', -0.9354838709677419)])

In [111]:
result = joinFreq.map(lambda e: (e[0], (e[1][0] - e[1][1])/(e[1][0] + e[1][1])))
result.cache()
result.takeOrdered(10, lambda v: -v[1]), result.takeOrdered(10, lambda v: +v[1])

([('pieza', 0.8),
  ('corral', 0.8),
  ('rodela', 0.7777777777777778),
  ('curar', 0.75),
  ('valle', 0.75),
  ('entierro', 0.75),
  ('oh', 0.7142857142857143),
  ('licor', 0.7142857142857143),
  ('difunto', 0.7142857142857143),
  ('pago', 0.6666666666666666)],
 [('teresa', -0.9767441860465116),
  ('roque', -0.96),
  ('paje', -0.9565217391304348),
  ('duque', -0.9565217391304348),
  ('blanca', -0.9565217391304348),
  ('gobernador', -0.9503105590062112),
  ('diego', -0.9459459459459459),
  ('tarde', -0.9428571428571428),
  ('mesmo', -0.9381443298969072),
  ('letras', -0.9354838709677419)])

La segunda. Ya que la primera calcula la lambda sobre el rdd para obtener result 2 veces. En la segunda cacheamos el resultado para queries que requieren de la misma transformación intermedia, haciendo más rapido la ejecución

In [112]:
result.coalesce(numPartitions=2) # Avoids the data movement, so it tries to balance inside each machine
result.repartition(numPartitions=2) # We don't care about data movement, this balance the whole thing to ensure all machines are used

MapPartitionsRDD[240] at coalesce at NativeMethodAccessorImpl.java:0

In [113]:
result.take(10)
allWords.cache() # allWords RDD must  stay in memory after computation, we made a checkpoint (well, it's a best effort, so must might be too strong)
result.take(10)

[('el', -0.5620334162815499),
 ('hidalgo', -0.5),
 ('don', -0.6255060728744939),
 ('mancha', -0.5905511811023622),
 ('saavedra', 0.0),
 ('que', -0.5361077111383109),
 ('y', -0.5789904278818621),
 ('del', -0.5281409891984082),
 ('en', -0.5704722945332837),
 ('cuyo', -0.5217391304347826)]

In [114]:
from pyspark import StorageLevel
# https://spark.apache.org/docs/2.2.0/rdd-programming-guide.html#rdd-persistence
allWords2.persist(StorageLevel.MEMORY_AND_DISK) # Now it will be preserved on disk also

PythonRDD[198] at RDD at PythonRDD.scala:53

In [120]:
!rm -rf palabras_parte2
#allWords2.saveAsTextFile("palabras_parte2")

allWords2.coalesce(numPartitions=6).saveAsTextFile("palabras_parte2")


In [121]:

%ls -la /content/palabras_parte2/

total 1100
drwxr-xr-x 2 root root   4096 Oct  5 08:01 ./
drwxr-xr-x 1 root root   4096 Oct  5 08:01 ../
-rw-r--r-- 1 root root 666813 Oct  5 08:01 part-00000
-rw-r--r-- 1 root root   5220 Oct  5 08:01 .part-00000.crc
-rw-r--r-- 1 root root 430302 Oct  5 08:01 part-00001
-rw-r--r-- 1 root root   3372 Oct  5 08:01 .part-00001.crc
-rw-r--r-- 1 root root      0 Oct  5 08:01 _SUCCESS
-rw-r--r-- 1 root root      8 Oct  5 08:01 ._SUCCESS.crc


## Pregunta TS2.6 Antes de guardar el fichero, utilice coalesce con diferente valores ¿Cuál es la diferencia?

Que reduce el número de particiones del rdd cuando lo guarda. Lo podemos ver haciendo un ls al directorio(ver codigo de arriba). Cabe destacar que como el número de particiones es originalmente 2, si ponemos un 1 lo reduce a 1 particion, 2 lo mantiene, y valores superiores lo mantiene también(puesto que solo reduce el numero de particiones). Logicamente el valor 0 no es aceptado.

Variables Globales

In [101]:
articles = sc.broadcast(["el", "la"])
articles.value

['el', 'la']

In [102]:
acc = sc.accumulator(0)
def incrementar(x):
  global acc
  acc += x

allWords.map(lambda l:1).foreach(incrementar)
acc

Accumulator<id=0, value=56521>

In [103]:
print (rdd.getStorageLevel())
rdd = sc.textFile('elquijote.txt')
%time print (rdd.count())
%time print (rdd.count())
rdd.cache()
%time print (rdd.count())
%time print (rdd.count())


Serialized 1x Replicated
5534
CPU times: user 9.44 ms, sys: 1.21 ms, total: 10.7 ms
Wall time: 94.9 ms
5534
CPU times: user 12.7 ms, sys: 0 ns, total: 12.7 ms
Wall time: 112 ms
5534
CPU times: user 9.69 ms, sys: 787 µs, total: 10.5 ms
Wall time: 150 ms
5534
CPU times: user 16.1 ms, sys: 780 µs, total: 16.8 ms
Wall time: 154 ms


In [104]:
pals_a_eliminar = ['a', 'ante', 'bajo', 'segun', 'que', 'de']

def elimPalabras(p):
    global pals_a_eliminar
    return p not in pals_a_eliminar

lineas = sc.textFile('elquijote.txt', 8)
pals = (lineas.flatMap(lambda linea: linea.lower().split())
              .filter(elimPalabras)
              .map(lambda pal: (pal, 1))
              .reduceByKey(lambda elem1,elem2: elem1 + elem2))

print (pals.takeOrdered(15, key=lambda a:-a[1]))

[('y', 2571), ('la', 1423), ('el', 1224), ('en', 1155), ('no', 880), ('se', 753), ('los', 696), ('con', 621), ('su', 617), ('por', 573), ('le', 534), ('las', 514), ('lo', 508), ('del', 415), ('don', 368)]


In [105]:
counter = sc.accumulator(0)

rdd = sc.textFile('elquijote.txt')

def increment_counter(x):
    global counter
    counter += x
    
rdd.map(lambda l:len(l)).foreach(increment_counter)

print ("Counter value: %d" % counter.value)

Counter value: 305678
